include/hipSYCL/sycl/libkernel/vec.hpp

```diff
@@ -182,6 +182,21 @@ template <class Vector_type, class Function>
 HIPSYCL_UNIVERSAL_TARGET
 void for_each_vector_element(const Vector_type& v, Function&& f);
 
+template <typename Arg, typename T, typename = void,
+          typename = std::enable_if_t<std::is_scalar_v<T>>>
+struct get_size {
+  static constexpr int size = 1;
+};
+
+template <typename Arg, typename T>
+struct get_size<Arg, T,
+                std::void_t<decltype(Arg::size())>,
+                std::enable_if_t<std::is_same_v<typename Arg::element_type, T>>> {
+  static constexpr int size = Arg::size();
+};
+
+template <typename Arg, typename T>
+constexpr int count_num_elements = get_size<Arg, T>::size;
```

Comment:
Is a `static constexpr` member function not sufficient here?

Reply:
For some reason, I couldn't make it work with a `static constexpr` member function (and I don't really understand why, that's one reason why I opened it as a draft PR for now) but I'll check that again.

Reply:
A `static constexpr` function would work but not with the `if constexpr` as before, for a similar reason as in my other comment: when both branches of a `if constexpr` inside a function that is used in a template parameter fail, then this is not just a substitution failure but will cause a compilation error, so you can't really use `if constexpr` for SFINAE (at least that's how I understand it, I'm not really 100% sure).